spec/rspec/core/reentrant_mutex_spec.rb

```diff
@@ -0,0 +1,37 @@
+require 'rspec/core/reentrant_mutex'
+
+# There are no assertions specifically
+# They are pass if they don't deadlock
+RSpec.describe RSpec::Core::ReentrantMutex do
+  let!(:mutex) { described_class.new }
+  let!(:order) { ThreadOrderSupport.new }
+  after { order.apocalypse! }
+
+  it 'can repeatedly synchronize within the same thread' do
+    mutex.synchronize { mutex.synchronize { } }
+  end
+
+  it 'locks other threads out while in the synchronize block' do
+    order.declare(:before) { mutex.synchronize { } }
+    order.declare(:within) { mutex.synchronize { } }
+    order.declare(:after)  { mutex.synchronize { } }
+
+    order.pass_to :before, :resume_on => :exit
+    mutex.synchronize { order.pass_to :before, :resume_on => :sleep }
+    order.pass_to :before, :resume_on => :exit
+  end
+
+  it 'resumes the next thread once all its synchronize blocks have completed' do
+    order.declare(:thread) { mutex.synchronize { } }
+    mutex.synchronize { order.pass_to :thread, :resume_on => :sleep }
+    order.apocalypse! :join
+  end
+
+  it 'is implemented without depending on the stdlib' do
+    loaded_filenames = $LOADED_FEATURES.map { |filepath| File.basename filepath }
+    pending 'thread seems to be required from core, and something is still requiring monitor'
+    expect(loaded_filenames).to_not include 'monitor.rb'
+    expect(loaded_filenames).to_not include 'thread.rb'
+    expect(loaded_filenames).to_not include 'thread.bundle'
+  end
```

Comment:
Yeah, I saw the rake one, commented on it in the "Segregating env from stdlib" section [here](https://github.com/rspec/rspec-core/pull/1858#issuecomment-75694930). I'm mostly uncomfortable with it loading because it reduces my confidence that this code is actually independent. I can try to either move these tests, or those tests out of the main process (think I saw the suite does this in a few places already). Of those two options, I'd prefer to move the rake and drb tests out, because then it means that we can guarantee all the specs in the main process run without those stdlib additions (ie segregate things that dirty the environment rather than things we want to make sure run in a clean environment). This seems worthwhile, I assume it's the motivation behind the "library wide checks". Though, I'm confused about how those pass (haven't looked into them yet) given that monitor and thread do get required.